internal/builtin/providers/tf/provider_functions.go

```diff
@@ -0,0 +1,173 @@
+	return "encode_tfvars"
+}
+
+func (f *encodeTFVarsFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name: "input",
+			// The input type is determined at runtime
+			Type:              cty.DynamicPseudoType,
+			Description:       "Input to encode for TFVars file. Must be an object with key that are valid identifiers",
+			DescriptionFormat: providers.TextFormattingPlain,
+		},
+	}
+	return providers.FunctionSpec{
+		Parameters:        params,
+		Return:            cty.String,
+		Summary:           "Encode an object into a string with the same format as a TFVars file",
+		Description:       "provider::terraform::encode_tfvars encodes an object into a string with the same format as a TFVars file",
+		DescriptionFormat: providers.TextFormattingPlain,
+	}
+}
+
+var InvalidInputError = errors.New("invalid input")
+
+func (f *encodeTFVarsFunc) Call(args []cty.Value) (cty.Value, error) {
+	toEncode := args[0]
+	// null is invalid input
+	if toEncode.IsNull() {
+		return cty.NullVal(cty.String), fmt.Errorf("%w: must not be null", InvalidInputError)
+	}
+	if !toEncode.Type().IsObjectType() {
+		return cty.NullVal(cty.String), fmt.Errorf("%w: must be an object", InvalidInputError)
+	}
+	ef := hclwrite.NewEmptyFile()
+	body := ef.Body()
+
+	// Iterate over the elements of the input value
+	it := toEncode.ElementIterator()
+	for it.Next() {
+		key, val := it.Element()
+		// Check if the key is a string, known and not null, otherwise AsString method panics
+		if !key.Type().Equals(cty.String) || !key.IsKnown() || key.IsNull() {
+			return cty.NullVal(cty.String), fmt.Errorf("%w: object key must be a string: %v", InvalidInputError, key)
+		}
+		name := key.AsString()
+		if valid := hclsyntax.ValidIdentifier(name); !valid {
+			return cty.NullVal(cty.String), fmt.Errorf("%w: object key: %s - must be a valid identifier", InvalidInputError, name)
+		}
+		body.SetAttributeValue(key.AsString(), val)
+	}
+	b := ef.Bytes()
+	return cty.StringVal(string(b)), nil
+}
+
+// encodeExprFunc encodes an expression into a string
+type encodeExprFunc struct{}
+
+func (f *encodeExprFunc) Name() string {
+	return "encode_expr"
+}
+
+func (f *encodeExprFunc) GetFunctionSpec() providers.FunctionSpec {
+	params := []providers.FunctionParameterSpec{
+		{
+			Name:              "expr",
+			Type:              cty.DynamicPseudoType,
+			Description:       "expression to encode",
+			DescriptionFormat: providers.TextFormattingPlain,
+		},
+	}
+	return providers.FunctionSpec{
+		Parameters:        params,
+		Return:            cty.String,
+		Summary:           "Takes an arbitrary expression and converts it into a string with valid OpenTofu syntax",
+		Description:       "provider::terraform::encode_expr takes an arbitrary expression and converts it into a string with valid OpenTofu syntax",
+		DescriptionFormat: providers.TextFormattingPlain,
+	}
+}
+
+var UnknownInputError = errors.New("input is not known")
```

Comment:
Not sure if that's enough from UX perspective. What do we usually output if the value is not wholly known?

Also, makes sense to double check other error messages (however, others look good to me).

Reply:
In a few places, we output that it is not wholly known, in other cases, we say that the value isn't known until the apply.
Do you have any suggestions on what we should output in this case?

Reply:
I think this is completely fine then to have an error message saying "input is not wholly known", it would result in something like:

```
│ Error: Error in function call
│ 
│   on main.tf line 10, in output "name":
│   10:   value = provider::terraform::encode_tfvars(var.something)
│     ├────────────────
│     │ while calling provider::terraform::encode_tfvars(input)
│ 
│ Call to function "provider::terraform::encode_tfvars" failed: input is not wholy known.
```